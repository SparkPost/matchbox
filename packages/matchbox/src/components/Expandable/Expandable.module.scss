@import '../../styles/config';

.Expandable {
  margin-bottom: spacing();
  border: 2px solid color(gray, 8);
  border-radius: border-radius(large);
}

.Header {
  display: flex;
  align-items: center;
  padding: rem(12);

  user-select: none;
  outline: none;

  transition: background 0.1s;

  &:hover {
    cursor: pointer;
  }

  &:focus:not(:hover) {
    background: color(gray, 9);
  }
}

.HeaderContent {
  flex: 1;
}

.HeaderArrow {
  flex: 0;
}

.Title {
  margin-bottom: 0;
}

.Subtitle {
  margin-bottom: 0;
  font-weight: normal;
}

.ContentWrapper {
  padding: rem(12);
  visibility: hidden;
  display: none;

  &.Open {
    visibility: visible;
    display: flex;
  }
}

.Content {
  flex: 1;
}

.ContentSpacer {
  flex: 0;
  margin-right: rem(12);
  min-width: 40px;
  max-width: 40px;
}

.Icon {
  flex: 0;
  margin-right: rem(12);
  min-width: 40px;
  max-width: 40px;

  svg {
    vertical-align: middle;
  }
}

.Arrow {
  border-radius: 50%;
  padding: spacing(small);
  transition: 0.1s ease-in-out;
  transform-origin: center;

  &.Open {
    transform: rotate(90deg);
  }

  .Header:hover & {
    background: color(gray, 9);
  }
}
