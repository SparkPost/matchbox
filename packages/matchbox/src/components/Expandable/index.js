export { default as Expandable } from './Expandable';
