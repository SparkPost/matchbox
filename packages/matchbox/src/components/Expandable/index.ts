export { default as Expandable } from './Expandable';
export * from './Expandable';
