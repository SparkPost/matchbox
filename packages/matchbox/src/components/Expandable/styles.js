import { tokens } from '@sparkpost/design-tokens';
import { buttonReset } from '../../styles/helpers';
import styled from 'styled-components';

export const StyledHeader = styled('button')`
  ${buttonReset}
  user-select: none;
  outline: none;
  padding: ${({ variant, theme }) => (variant === 'borderless' ? '0' : theme.space[300])};
  display: flex;
  align-items: center;
  text-align: left;
  width: 100%;

  transition: background ${tokens.motionDuration_fast} ${tokens.motionEase_in_out};

  &:hover {
    cursor: pointer;
  }

  &:focus:not(:hover) {
    background: ${tokens.color_gray_200};
  }
`;

export const expandable = props => {
  let borderRadius = tokens.borderRadius_100;
  if (props.variant === 'borderless') {
    return ``;
  }

  if (props.accent) {
    borderRadius = `0 0 ${tokens.borderRadius_100} ${tokens.borderRadius_100}`;
  }

  return `
    border: ${props.theme.borders['300']};
    border-radius: ${borderRadius};
    border-top-width: ${props.accent ? tokens.borderWidth_0 : tokens.borderWidth_100}
  `;
};

export const arrow = props => {
  let rotate = 'rotate(-90deg)';

  if (props.isOpen) {
    rotate = 'rotate(90deg)';
  }

  return `
    border-radius: ${tokens.borderRadius_circle};
    padding: ${tokens.spacing_100};
    transition: ${tokens.motionDuration_fast} ${tokens.motionEase_in_out};
    transform-origin: center;

    transform: ${rotate};

    ${StyledHeader}:hover & {
      background: ${tokens.color_gray_200};
    }
  `;
};

export const contentWrapper = props => {
  let visibility = 'hidden';
  let display = 'none';

  if (props.isOpen) {
    visibility = 'visible';
    display = 'flex';
  }

  return `
    padding: ${
      props.variant === 'borderless' ? `${props.theme.space[300]} 0 0` : props.theme.space[300]
    };
    visibility: ${visibility};
    display: ${display};
  `;
};

export const title = () => `
  font-size: ${tokens.fontSize_400};
  font-weight: ${tokens.fontWeight_semibold};
`;

export const subtitle = () => `
  font-weight: ${tokens.fontWeight_normal};
`;

export const accent = props => {
  let color;

  switch (props.accentColor) {
    case 'orange':
      color = tokens.color_brand_orange;
      break;
    case 'green':
      color = tokens.color_green_700;
      break;
    case 'yellow':
      color = tokens.color_yellow_400;
      break;
    case 'red':
      color = tokens.color_red_700;
      break;
    case 'gray':
      color = tokens.color_gray_600;
      break;
    case 'blue':
    case 'navy':
    case 'purple':
    default:
      color = tokens.color_blue_700;
  }

  return `
    height: ${tokens.spacing_100};
    border-radius: ${tokens.borderRadius_100} ${tokens.borderRadius_100} 0 0;
    background-color: ${color};
  `;
};
