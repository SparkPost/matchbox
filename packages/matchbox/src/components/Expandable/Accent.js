import React from 'react';

import styled from 'styled-components';
import PropTypes from 'prop-types';
import { accent } from './styles';

const AccentOuter = styled('div')`
  ${accent}
`;

const Accent = ({ accentColor }) => {
  return <AccentOuter accentColor={accentColor} />;
};

Accent.displayName = 'Expandable.Accent';
Accent.propTypes = {
  accentColor: PropTypes.oneOf([
    'orange',
    'blue',
    'red',
    'yellow',
    'green',
    'purple',
    'navy',
    'gray',
  ]),
};

export default Accent;
