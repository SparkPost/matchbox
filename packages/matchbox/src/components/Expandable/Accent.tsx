import React from 'react';
import styled from 'styled-components';
import { accent } from './styles';

const AccentOuter = styled('div')`
  ${accent}
`;

type ExpandableAccentProps = {
  accentColor?:
    | boolean
    | 'orange'
    | 'blue'
    | 'red'
    | 'yellow'
    | 'green'
    | 'purple'
    | 'navy'
    | 'gray';
};

const Accent = ({ accentColor }: ExpandableAccentProps): JSX.Element => {
  return <AccentOuter accentColor={accentColor} />;
};

Accent.displayName = 'Expandable.Accent';
export default Accent;
