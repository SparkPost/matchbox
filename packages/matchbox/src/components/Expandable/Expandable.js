import React from 'react';
import Proptypes from 'prop-types';
import styled from 'styled-components';
import { margin } from 'styled-system';
import { createPropTypes } from '@styled-system/prop-types';
import { KeyboardArrowRight } from '@sparkpost/matchbox-icons';
import { onKeys } from '../../helpers/keyEvents';
import { buttonReset } from '../../styles/helpers';
import { expandable, header, title, subtitle, arrow, contentWrapper } from './styles';

import Accent from './Accent';
import { Box } from '../Box';

const StyledExpandable = styled('div')`
  ${expandable}
  ${margin}
`;

export const StyledHeader = styled('button')`
  ${buttonReset}
  ${header}
`;

export const StyledArrow = styled('div')`
  ${arrow}
`;

export const StyledTitle = styled('h3')`
  ${title}
`;

const StyledSubtitle = styled('h6')`
  ${subtitle}
`;

const StyledContentWrapper = styled('div')`
  ${contentWrapper}
`;

const StyledIcon = styled(Box)``;

function Expandable(props) {
  const {
    children,
    defaultOpen,
    icon,
    id,
    open,
    onToggle,
    subtitle,
    title,
    accent,
    ...rest
  } = props;

  const header = React.useRef();
  const [isOpen, setIsOpen] = React.useState(open || defaultOpen);
  const controlled = typeof open === 'boolean';

  // Sets internal open state when controlled externally
  React.useEffect(() => {
    if (controlled) {
      setIsOpen(open);
    }
  }, [open]);

  function handleToggle() {
    if (controlled) {
      onToggle();
    } else {
      setIsOpen(!isOpen);
    }
  }

  function handleClick() {
    header.current.blur();
    handleToggle();
  }

  function handleKeyDown(e) {
    onKeys(['space', 'enter'], () => {
      e.preventDefault();
      handleToggle();
    })(e);
  }

  const accentColor = accent === true ? 'blue' : accent;

  const accentMarkup = accentColor ? <Accent accentColor={accentColor} /> : null;

  const iconMarkup = icon ? (
    <StyledIcon flex="0" minWidth="40px" maxWidth="40px" mr="500">
      {icon}
    </StyledIcon>
  ) : null;

  const subtitleMarkup = subtitle ? <StyledSubtitle>{subtitle}</StyledSubtitle> : null;

  const contentSpacer = icon ? <Box flex="0" minWidth="40px" maxWidth="40px" mr="500" /> : null;

  return (
    <Box {...rest}>
      {accentMarkup}
      <StyledExpandable accent={accent}>
        <StyledHeader
          aria-controls={id}
          aria-expanded={isOpen}
          onClick={handleClick}
          onKeyDown={handleKeyDown}
          ref={header}
          role="button"
          tabIndex="0"
          data-id="expandable-toggle"
        >
          {iconMarkup}
          <Box display="inline-block" flex="1">
            <StyledTitle>{title}</StyledTitle>
            {subtitleMarkup}
          </Box>
          <Box display="inline-block" flex="0">
            <StyledArrow isOpen={isOpen}>
              <KeyboardArrowRight size={26} />
            </StyledArrow>
          </Box>
        </StyledHeader>
        <StyledContentWrapper
          aria-hidden={!isOpen}
          isOpen={isOpen}
          id={id}
          data-id="expandable-content"
        >
          {contentSpacer}
          <Box flex="1">{children}</Box>
        </StyledContentWrapper>
      </StyledExpandable>
    </Box>
  );
}

Expandable.Accent = Accent;
Expandable.Icon = StyledIcon;
Expandable.Title = StyledTitle;
Expandable.Subtitle = StyledSubtitle;
Expandable.ContentWrapper = StyledContentWrapper;
Expandable.Arrow = StyledArrow;
Expandable.Header = StyledHeader;

Expandable.defaultProps = {
  defaultOpen: false,
};

Expandable.propTypes = {
  /**
   * Default open state when not controlled
   */
  defaultOpen: Proptypes.bool,
  /**
   * Optional React node for an image or icon. Works best with an SVG optimized for 40x40.
   */
  icon: Proptypes.node,
  /**
   * Required for accessilibty between header and content
   */
  id: Proptypes.string.isRequired,
  /**
   * Pass a boolean to control open state
   */
  open: Proptypes.bool,
  /**
   * Optional, but required when controlling open state
   */
  onToggle: Proptypes.func,
  /**
   * Optional content area beneath header title
   */
  subtitle: Proptypes.node,
  /**
   * Main header title content
   */
  title: Proptypes.node.isRequired,
  /**
   * Optional accent color or boolean to default to blue
   */
  accent: Proptypes.oneOfType([
    Proptypes.bool,
    Proptypes.oneOf(['orange', 'blue', 'red', 'yellow', 'green', 'purple', 'navy', 'gray']),
  ]),
  /**
   * Margin props
   */
  ...createPropTypes(margin.propNames),
};

export default Expandable;
