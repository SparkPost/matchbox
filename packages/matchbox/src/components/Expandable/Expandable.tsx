import React from 'react';
import styled from 'styled-components';
import { margin, MarginProps } from 'styled-system';
import { KeyboardArrowLeft } from '@sparkpost/matchbox-icons';
import { onKeys } from '../../helpers/keyEvents';
import { pick } from '../../helpers/props';
import { StyledHeader, StyledContentWrapper, expandable, title, subtitle, arrow } from './styles';

import Accent from './Accent';
import { Box, BoxProps } from '../Box';

type TransientProps = {
  $accent?: React.ComponentProps<typeof Accent>['accentColor'];
  $variant?: string;
  $isOpen?: boolean;
};

const StyledExpandable = styled('div')<TransientProps>`
  ${expandable}
  ${margin}
`;

const StyledArrow = styled.div<TransientProps>`
  ${arrow}
`;

const StyledTitle = styled('h3')<TransientProps>`
  ${title}
`;

const StyledSubtitle = styled(Box)<BoxProps & TransientProps>`
  ${subtitle}
`;

export type ExpandableProps = MarginProps & {
  /**
   * Optional accent color or boolean to default to blue
   */
  accent?: React.ComponentProps<typeof Accent>['accentColor'];
  /**
   * Default open state when not controlled
   */
  defaultOpen?: boolean;
  /**
   * Optional React node for an image or icon. Works best with an SVG optimized for 40x40.
   */
  icon?: React.ReactNode;
  /**
   * Required for accessilibty between header and content
   */
  id: string;
  /**
   * Pass a boolean to control open state
   */
  open?: boolean;
  /**
   * Optional, but required when controlling open state
   */
  onToggle?: () => void;
  /**
   * Optional content area beneath header title
   */
  subtitle?: React.ReactNode;
  /**
   * Main header title content
   */
  title: React.ReactNode;
  /**
   * Contents of the expandable
   */
  children?: React.ReactNode;
  'data-id'?: string;
  variant?: 'bordered' | 'borderless';
};

function Expandable(props: ExpandableProps): JSX.Element {
  const {
    children,
    defaultOpen = false,
    'data-id': dataId,
    icon,
    id,
    open,
    onToggle,
    subtitle,
    title,
    accent,
    variant = 'bordered',
    ...rest
  } = props;

  const systemProps = pick(rest, margin.propNames);
  const header = React.useRef();
  const [isOpen, setIsOpen] = React.useState(open || defaultOpen);
  const controlled = typeof open === 'boolean';

  // Sets internal open state when controlled externally
  React.useEffect(() => {
    if (controlled) {
      setIsOpen(open);
    }
  }, [open]);

  function handleToggle() {
    if (controlled) {
      onToggle();
    } else {
      setIsOpen(!isOpen);
    }
  }

  function handleClick() {
    (header as React.MutableRefObject<HTMLButtonElement>).current.blur();
    handleToggle();
  }

  function handleKeyDown(e) {
    onKeys(['space', 'enter'], () => {
      e.preventDefault();
      handleToggle();
    })(e);
  }

  const accentColor = accent === true ? 'blue' : accent;

  const accentMarkup = accentColor ? <Accent accentColor={accentColor} /> : null;

  const iconMarkup = icon ? (
    <Box flex="0" minWidth="40px" maxWidth="40px" mr="500">
      {icon}
    </Box>
  ) : null;

  const subtitleMarkup = subtitle ? (
    <StyledSubtitle as="h6" mb="100">
      {subtitle}
    </StyledSubtitle>
  ) : null;

  const contentSpacer = icon ? <Box flex="0" minWidth="40px" maxWidth="40px" mr="500" /> : null;

  return (
    <Box data-id={dataId} {...systemProps}>
      {accentMarkup}
      <StyledExpandable $accent={accent} $variant={variant}>
        <StyledHeader
          aria-controls={id}
          aria-expanded={isOpen}
          onClick={handleClick}
          onKeyDown={handleKeyDown}
          ref={header}
          data-id="expandable-toggle"
          type="button"
          $variant={variant}
        >
          {iconMarkup}
          <Box display="inline-block" flex="1">
            <StyledTitle $variant={variant}>{title}</StyledTitle>
            {subtitleMarkup}
          </Box>
          <Box display="inline-block" flex="0">
            <StyledArrow $isOpen={isOpen}>
              <KeyboardArrowLeft size={26} />
            </StyledArrow>
          </Box>
        </StyledHeader>
        <StyledContentWrapper
          aria-hidden={!isOpen}
          $isOpen={isOpen}
          id={id}
          data-id="expandable-content"
          $variant={variant}
        >
          {contentSpacer}
          <Box flex="1">{children}</Box>
        </StyledContentWrapper>
      </StyledExpandable>
    </Box>
  );
}

Expandable.Accent = Accent;
Expandable.Title = StyledTitle;
Expandable.Subtitle = StyledSubtitle;
Expandable.ContentWrapper = StyledContentWrapper;
Expandable.Arrow = StyledArrow;
Expandable.Header = StyledHeader;
export default Expandable;
