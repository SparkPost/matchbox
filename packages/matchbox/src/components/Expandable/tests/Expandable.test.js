import React from 'react';
import Expandable from '../Expandable';
import { mount } from 'enzyme';

describe('Expandable component', () => {
  let onToggle;

  beforeEach(() => {
    onToggle = jest.fn();
  });

  const testOpen = (wrapper, open) => {
    expect(wrapper.find('.Header')).toHaveAttributeValue('aria-expanded', open ? 'true' : 'false');
    expect(wrapper.find('svg.Arrow')).toHaveAttributeValue('class', open ? 'Arrow Open' : 'Arrow');
    expect(wrapper.find('.ContentWrapper')).toHaveAttributeValue('class', open ? 'ContentWrapper Open' : 'ContentWrapper');
    expect(wrapper.find('.ContentWrapper')).toHaveAttributeValue('aria-hidden', open ? 'false' : 'true');
  };

  const subject = (props = {}) => mount(
    <Expandable title='test-title' id='test-id' {...props}/>
  );

  it('should render default state with title and id correctly', () => {
    expect(subject()).toMatchSnapshot();
  });

  it('should render provided default state with an icon', () => {
    expect(subject({
      icon: <svg>test</svg>,
      defaultOpen: true
    })).toMatchSnapshot();
  });

  describe('uncontrolled', () => {
    it('should use default defaultOpen on first render', () => {
      const wrapper = subject();
      testOpen(wrapper, false);
    });

    it('should use defaultOpen on first render', () => {
      const wrapper = subject({ defaultOpen: true });
      testOpen(wrapper, true);
    });

    it('should toggle open when clicking on header', () => {
      const wrapper = subject();
      wrapper.find('.Header').simulate('click');
      testOpen(wrapper, true);
    });

    it('should toggle when hitting enter or space', () => {
      const wrapper = subject({ onToggle });
      wrapper.find('.Header').simulate('keyDown', { key: 'Enter', shiftKey: false });
      testOpen(wrapper, true);

      wrapper.find('.Header').simulate('keyDown', { key: ' ', shiftKey: false });
      testOpen(wrapper, false);
    });
  });

  describe('controlled', () => {
    it('should toggle open when passed open state', () => {
      const wrapper = subject({ open: true, onToggle });
      testOpen(wrapper, true);
    });

    it('should handle toggle when clicking on header', () => {
      const wrapper = subject({ open: true, onToggle });
      wrapper.find('.Header').simulate('click');
      expect(onToggle).toHaveBeenCalled();
    });

    it('should handle toggle when hitting enter or space', () => {
      const wrapper = subject({ open: true, onToggle });
      wrapper.find('.Header').simulate('keyDown', { key: 'Enter', shiftKey: false });
      wrapper.find('.Header').simulate('keyDown', { key: ' ', shiftKey: false });
      expect(onToggle).toHaveBeenCalledTimes(2);
    });
  });
});
