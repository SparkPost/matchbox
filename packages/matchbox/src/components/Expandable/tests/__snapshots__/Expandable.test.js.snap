// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expandable component should render default state with title and id correctly 1`] = `
<Expandable
  defaultOpen={false}
  id="test-id"
  title="test-title"
>
  <div
    className="Expandable"
  >
    <div
      aria-controls="test-id"
      aria-expanded={false}
      className="Header"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex="1"
    >
      <div
        className="HeaderContent"
      >
        <h3
          className="Title"
        >
          test-title
        </h3>
        <h6
          className="Subtitle"
        />
      </div>
      <div
        className="HeaderArrow"
      >
        <KeyboardArrowRight
          className="Arrow"
          size={36}
        >
          <IconBase
            className="Arrow"
            size={36}
          >
            <svg
              aria-hidden="true"
              className="Arrow"
              fill="currentColor"
              height={36}
              preserveAspectRatio="xMidYMid meet"
              style={
                Object {
                  "verticalAlign": "middle",
                }
              }
              viewBox="0 0 24 24"
              width={36}
            >
              <g>
                <path
                  d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
                />
              </g>
            </svg>
          </IconBase>
        </KeyboardArrowRight>
      </div>
    </div>
    <div
      aria-hidden={true}
      className="ContentWrapper"
      id="test-id"
    >
      <div
        className="Content"
      />
    </div>
  </div>
</Expandable>
`;

exports[`Expandable component should render provided default state with an icon 1`] = `
<Expandable
  defaultOpen={true}
  icon={
    <svg>
      test
    </svg>
  }
  id="test-id"
  title="test-title"
>
  <div
    className="Expandable"
  >
    <div
      aria-controls="test-id"
      aria-expanded={true}
      className="Header"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex="1"
    >
      <div
        className="Icon"
      >
        <svg>
          test
        </svg>
      </div>
      <div
        className="HeaderContent"
      >
        <h3
          className="Title"
        >
          test-title
        </h3>
        <h6
          className="Subtitle"
        />
      </div>
      <div
        className="HeaderArrow"
      >
        <KeyboardArrowRight
          className="Arrow Open"
          size={36}
        >
          <IconBase
            className="Arrow Open"
            size={36}
          >
            <svg
              aria-hidden="true"
              className="Arrow Open"
              fill="currentColor"
              height={36}
              preserveAspectRatio="xMidYMid meet"
              style={
                Object {
                  "verticalAlign": "middle",
                }
              }
              viewBox="0 0 24 24"
              width={36}
            >
              <g>
                <path
                  d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
                />
              </g>
            </svg>
          </IconBase>
        </KeyboardArrowRight>
      </div>
    </div>
    <div
      aria-hidden={false}
      className="ContentWrapper Open"
      id="test-id"
    >
      <div
        className="ContentSpacer"
      />
      <div
        className="Content"
      />
    </div>
  </div>
</Expandable>
`;
